cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextProvider.java

```diff
@@ -16,57 +16,203 @@
+import co.cask.cdap.common.guice.DiscoveryRuntimeModule;
+import co.cask.cdap.common.guice.IOModule;
+import co.cask.cdap.common.guice.KafkaClientModule;
+import co.cask.cdap.common.guice.LocationRuntimeModule;
+import co.cask.cdap.common.guice.ZKClientModule;
+import co.cask.cdap.common.lang.ClassLoaders;
+import co.cask.cdap.common.lang.ProgramClassLoader;
+import co.cask.cdap.data.runtime.DataFabricModules;
+import co.cask.cdap.data.runtime.DataSetsModules;
+import co.cask.cdap.data.stream.StreamAdminModules;
+import co.cask.cdap.data.stream.StreamCoordinatorClient;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.transaction.stream.StreamAdmin;
+import co.cask.cdap.logging.appender.LogAppenderInitializer;
+import co.cask.cdap.logging.guice.LoggingModules;
+import co.cask.cdap.metrics.guice.MetricsClientRuntimeModule;
+import co.cask.cdap.notifications.feeds.guice.NotificationFeedServiceRuntimeModule;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.AbstractService;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.Service;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.mapreduce.MRConfig;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.internal.Services;
+import org.apache.twill.kafka.client.KafkaClientService;
+import org.apache.twill.zookeeper.ZKClientService;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
 
 /**
- * Provides access to SparkContext for Spark job tasks
+ * Helper class for locating {@link ExecutionSparkContext} from the execution context.
  */
-public class SparkContextProvider {
+public final class SparkContextProvider {
+
+  private static volatile ExecutionSparkContext sparkContext;
 
-  private final Configuration hConf;
-  private final SparkContextConfig contextConfig;
-  private BasicSparkContext context;
-  private AbstractSparkContextBuilder contextBuilder;
+  /**
+   * Returns the current {@link ExecutionSparkContext}.
+   */
+  public static ExecutionSparkContext getSparkContext() {
+    if (sparkContext != null) {
+      return sparkContext;
+    }
 
-  public SparkContextProvider(Configuration hConf) {
-    this.hConf = hConf;
-    this.contextConfig = new SparkContextConfig(hConf);
-    this.contextBuilder = null;
+    // Try to find it from the context classloader
+    SparkClassLoader sparkClassLoader = ClassLoaders.find(Thread.currentThread().getContextClassLoader(),
+                                                          SparkClassLoader.class);
+    if (sparkClassLoader != null) {
+      return sparkClassLoader.getContext();
```

Comment:
would you want to set sparkContext=sparkClassLoader.getContext(); return sparkContext;